Retire NotificationPresenter in favor of view components

saraycp opened the pull request:
`NotificationComponent` renders a notification with the help of a few small view components, each having a single responsibility.


Fixes #11558

danidoni commented on src/api/app/components/notification_avatars_component.html.haml:
```diff
@@ -0,0 +1,16 @@
+%ul.list-inline.d-flex.flex-row-reverse.avatars.m-0
+  - if avatar_objects.size > 6
+    %li.list-inline-item
+      - hidden = (avatar_objects.size - 6)
```
What about moving this logic into the component? That way you can document those `6` and present a simpler interface in the view. :thought_balloon:

saraycp replied:
I added a `NotificationAvatarsComponent` for this. I also refactored the code a bit, moving logic from the view to the component and using a constant instead of `6`, among others.

danidoni commented on src/api/app/views/webui/users/notifications/_notifications_list.html.haml:
```diff
@@ -21,46 +21,8 @@
-                  .row.d-none.d-md-block.pl-4
-                    .col
-                      %p.mt-3.mb-0= notification.excerpt
+          - notifications.each do |notification|
+            = render NotificationComponent.new(notification, selected_filter)
```
I think you can use the collections helpers in view controllers[^1] to render collections.

[^1]: https://viewcomponent.org/guide/collections.html

saraycp replied:
True! I've changed the call to the component to use the collection instead.

danidoni commented on src/api/app/components/notification_component.html.haml:
```diff
@@ -0,0 +1,40 @@
+            %small.text-nowrap #{time_ago_in_words(@notification.created_at)} ago
+        .col-auto.actions.ml-auto.align-self-end.align-self-md-start
+          :ruby
+            title, icon = @notification.unread? ? ['Mark as "Read"', 'fa-check'] : ['Mark as "Unread"', 'fa-undo']
```
I think this can also be pushed up into the component so we only ask for the title and the icon in the view, and the component does the logic :thought_balloon:

saraycp replied:
I added a `NotificationMarkButtonComponent` for this.

saraycp commented:
@danidoni I addressed your feedback, can you have a look? 
I've created a `NotificationComponent` that calls to other components for many elements of the notification.

danidoni commented:
> @danidoni I addressed your feedback, can you have a look? I've created a `NotificationComponent` that calls to other components for many elements of the notification.

LGTM!

hennevogel commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Turn NotificationPresenter into a view component](https://trello.com/c/U8c9ghnP/2384-turn-notificationpresenter-into-a-view-component)
